### PR TITLE
Model object should support copy and deepcopy

### PR DESCRIPTION
**Pull request by krizex**
Enable the `copy` and `deepcopy` feature of model objects.
This PR is the fix for that I mentioned in #2483 .

### CODE REVIEW

**Comment by krizex** [comment]
Per https://docs.python.org/2/library/copy.html?highlight=copy.copy#copy.copy we can see:

> Classes can use the same interfaces to control copying that they use to control pickling. See the description of module pickle for information on these methods. The copy module does not use the copy_reg registration module.

Inspect to the source code, `copy.copy` and `copy.deepcopy` will call `_reconstruct` and in this method it will call the `__setstate__` method if the instance has it. When we implement `__setstate__` we should consider the argument as a tuple as described on https://www.python.org/dev/peps/pep-0307/ 

> Case 3: pickling new-style class instances using protocol 2

**Comment by pekkaklarck** [comment]
Did you investigate implementing `__copy__` and `__deepcopy__` instead of using the somewhat magical looking `__setstate__`?

**Comment by pekkaklarck** [comment]
It's good to see a reference to `pickle` that is documented to use `__setstate__` in `copy` module docs. And supporting `pickle` could be useful in some cases too.

**Comment by krizex** [comment]
Per the implementation of `copy.copy` and `copy.deepcopy`, we don't have to implement methods of `__copy__` and `__deepcopy__`, and we just need to implement the `__setstate__` as I did in this PR.

File: src/robot/model/modelobject.py [hunk 1]
```diff
@@ -26,4 +26,33 @@ def __unicode__(self):
...
+        """Customize the attribute updating when using `copy.copy` or `copy.deepcopy`
+        """
+
+        # We should consider the state format per pickle protocol version
+        # Refer to: https://www.python.org/dev/peps/pep-0307
+        if isinstance(state, tuple) and len(state) == 2:
+            dictstate, slotstate = state
+        else:
+            dictstate = state
+            slotstate = None
```
Comment (pekkaklarck) [inline src/robot/model/modelobject.py]: We have `__slots__` and can thus know, based on the referred PEP, that state is two-tuple.

File: src/robot/model/modelobject.py [hunk 2]
```diff
@@ -26,4 +26,33 @@ def __unicode__(self):
...
+        def get_setter_name(x):
+            return '_setter__' + x
+
+        if slotstate is not None:
+            for k, v in slotstate.iteritems():
+                setter_name = get_setter_name(k)
+                # If we have defined the attribute both in __slots__ and with @setter,
+                # we just need to set the setter attribute.
+                if setter_name in slotstate:
+                    continue
```
Comment (pekkaklarck) [inline src/robot/model/modelobject.py]: Could you clarify why is this needed? I tested that all unit tests pass also without it.

File: src/robot/model/modelobject.py [hunk 3]
```diff
@@ -26,4 +26,33 @@ def __unicode__(self):
...
+            dictstate, slotstate = state
+        else:
+            dictstate = state
+            slotstate = None
+
+        if dictstate is not None:
+            self.__dict__.update(dictstate)
+
+        def get_setter_name(x):
+            return '_setter__' + x
```
Comment (pekkaklarck) [inline src/robot/model/modelobject.py]: This function isn't really needed.

File: src/robot/model/modelobject.py [hunk 4]
```diff
@@ -26,4 +26,33 @@ def __unicode__(self):
...
+            for k, v in slotstate.iteritems():
+                setter_name = get_setter_name(k)
+                # If we have defined the attribute both in __slots__ and with @setter,
+                # we just need to set the setter attribute.
+                if setter_name in slotstate:
+                    continue
+
+                setattr(self, k, v)
+
+
```
Comment (pekkaklarck) [inline src/robot/model/modelobject.py]: Our code conventions say empty rows shouldn't generally be used inside functions/methods and there shouldn't be empty rows at the end.

File: utest/model/test_keyword.py [hunk 5]
```diff
@@ -33,6 +34,22 @@ def test_id_with_keyword_parents(self):
...
 
+    def test_copy_keyword(self):
+        kw = Keyword()
+        kw_new = copy.copy(kw)
+        self.assertEqual(kw.name, kw_new.name)
+
+        kw_new.name = kw.name + '1'
+        self.assertNotEqual(kw.name, kw_new.name)
+
+        self.assertEqual(id(kw.tags), id(kw_new.tags))
```
Comment (pekkaklarck) [inline utest/model/test_keyword.py]: Good tests. Two nitpick comments:

- We have our own `assert_equal` and friends.
- We don't use empty lines inside functions (unless functions are long but we also avoid long functions).

**Comment by pekkaklarck** [comment]
OK, I think I figured out why that `if setter_name in slotstate: continue` thing is needed: Without it copying a `running.TestCase` that has a timeout fails. There were some unit tests for copying these tests, but those tests didn't have timeout so unit tests passed also without the guarding if. I try to enhance tests so that they'll detect this problem too.

**Comment by pekkaklarck** [comment]
One more thing, do you @krizex want to be credited in `AUTHORS.txt`? Please add yourself there if you want to. You definitely deserve it after this PR.

**Comment by krizex** [comment]
@pekkaklarck , thanks for your kindly reminder about the code convention of this project. I have created a PR #2511  to modify the AUTHORS.txt, please merge it at your convenience.
Happy new year!

**Comment by pekkaklarck** [comment]
Thank you for the PR! Not sure had I found `__setstate__` myself or learned how to use it.
